hail/src/main/scala/is/hail/methods/MatrixExportColsParallel.scala

```diff
@@ -0,0 +1,143 @@
+          new OutputStreamWriter(hConfBc.value.value.unsafeWriter(partFolder + "/" + j.toString + extension), "UTF-8")
+        }
+
+        if (i == 0) {
+          // write headers
+          val header = (
+            rvType.fieldNames.filter(_ != MatrixType.entriesIdentifier) ++ entryType.fieldNames
+            ).mkString("\t")
+
+          fileHandles.zipWithIndex.foreach { case (f, jj) =>
+            f.write('#')
+            f.write(colValuesJSON.value(jj))
+            f.write('\n')
+            f.write(header)
+            f.write('\n')
+          }
+        }
+
+        it.foreach { rv =>
+
+          val entriesArray = new UnsafeIndexedSeq(entryArrayType, rv.region, rvType.loadField(rv, entriesIdx))
+
+          val fullRow = new UnsafeRow(rvType, rv)
+
+          val rowFieldStrs = (0 until rvType.size)
+            .filter(_ != entriesIdx)
+            .map { rowFieldIdx =>
+              TableAnnotationImpex.exportAnnotation(fullRow(rowFieldIdx), rvType.types(rowFieldIdx).virtualType)
+            }.toArray
+          fileHandles.indices.foreach { fileIdx =>
+            val entryIdx = fileIdx + startIdx
+            val os = fileHandles(fileIdx)
+
+            rowFieldStrs.foreach { s =>
+              os.write(s)
+              os.write('\t')
+            }
+
+            entriesArray(entryIdx) match {
+              case null =>
+                (0 until entryType.size).foreachBetween { _ =>
+                  os.write("NA")
+                }(os.write('\t'))
+              case r: Row =>
+                (0 until entryType.size).foreachBetween { entryFieldIdx =>
+                  os.write(TableAnnotationImpex.exportAnnotation(r.get(entryFieldIdx), entryType.types(entryFieldIdx).virtualType))
+                }(os.write('\t'))
+            }
+
+            os.write('\n')
+          }
+        }
+
+        fileHandles.foreach(_.close())
+        Iterator(partFolder)
+      }.collect()
+
+      val ns = endIdx - startIdx
+      val newFiles = mv.sparkContext.parallelize(0 until ns, numSlices = ns)
+        .map { sampleIdx =>
+          val partFilePath = path + "/" + partFile(digitsNeeded(nCols), sampleIdx, TaskContext.get)
+          val fileStatuses = partFolders.map(pf => hConfBc.value.value.fileStatus(pf + s"/$sampleIdx" + extension))
+          hConfBc.value.value.copyMergeList(fileStatuses, partFilePath, deleteSource = true)
+          partFilePath
+        }.collect()
+
+      newFiles
+    }
+
+    val padding = digitsNeeded(mv.nCols)
+    val extension = if (bgzip) ".tsv.bgz" else ".tsv"
+
+    def finalPath(idx: Int): String = path + "/" + partFile(padding, idx) + extension
+
+    resultFiles.zipWithIndex.foreach { case (filePath, i) =>
+      hConf.copy(filePath, finalPath(i), deleteSource = true)
+    }
+    hConf.delete(path + "/tmp", recursive = true)
+
+    hConf.writeTable(path + "/index.tsv", allColValuesJSON.zipWithIndex.map { case (json, i) =>
```

Comment:
Rather than exporting the column annotations as JSON, why not export it as a TSV with a new column which is the file path? Then this can be imported easily back into Hail.

Reply:
> export it

What is "it" in this case?

This format can easily be read back into Hail, no?

Reply:
sorry the column annotations. I'm asking why not generate what would be the input format for `import_table`

Reply:
I think that's exactly what I've done! Export a two-column file with 1) the file path, and b) the column struct as json